mlflow/utils/autologging_utils/__init__.py

```diff
@@ -514,46 +514,75 @@ def _get_new_training_session_class():
 
-            :param clazz: A class object that this session originates from.
+            :param estimator: An estimator that this session originates from.
             :param allow_children: If True, allows autologging in child sessions.
                                    If False, disallows autologging in all descendant sessions.
+                                   If current estimator is parent estimator (reenterring case),
+                                   this param is ignored.
             """
-            self.allow_children = allow_children
-            self.clazz = clazz
-            self._parent = None
+            current_session = _TrainingSession.get_current_session()
+            if current_session is not None and current_session.estimator is estimator:
+                current_session.reentrant_count += 1
+                return current_session
+            else:
+                if current_session is not None:
+                    allow_children = allow_children and current_session.allow_children
+                    new_session = _TrainingSession(
+                        estimator, allow_children, parent=current_session
+                    )
+                else:
+                    new_session = _TrainingSession(estimator, allow_children)
+                _TrainingSession._session_stack.append(new_session)
+                return new_session
 
         def __enter__(self):
-            if len(_TrainingSession._session_stack) > 0:
-                self._parent = _TrainingSession._session_stack[-1]
-                self.allow_children = (
-                    _TrainingSession._session_stack[-1].allow_children and self.allow_children
-                )
-            _TrainingSession._session_stack.append(self)
             return self
 
         def __exit__(self, tp, val, traceback):
-            _TrainingSession._session_stack.pop()
+            if self.reentrant_count > 0:
+                self.reentrant_count -= 1
+            else:
+                _TrainingSession._session_stack.pop()
 
         def should_log(self):
             """
-            Returns True when at least one of the following conditions satisfies:
+            Returns True when reentrant_count == 0 and at least one of the following conditions
+            satisfies:
 
             1. This session is the root session.
-            2. The parent session allows autologging and its class differs from this session's
-               class.
+            2. The parent session allows autologging and its estimator differs from this session's
+               differs.
             """
-            return (self._parent is None) or (
-                self._parent.allow_children and self._parent.clazz != self.clazz
-            )
+            return self.reentrant_count == 0 and self.should_log_without_reentrance()
+
+        def should_log_without_reentrance(self):
+            return self._parent is None or self._parent.allow_children
```

Comment:
@WeichenXu123 Instead of splitting into two methods and keeping a reentrant count, what if we update `should_log` to look through the entries in `_session_stack` above the current entry and, if there's an entry whose estimator matches the current instance, call `should_log()` on that entry? Would that work?